exercises/pythagorean-triplet/example.py

```diff
@@ -1,37 +1,61 @@
-def triplets_in_range(min, max):
-    result = set([x for b in
-                  range(4, max + 1, 4) for x in
-                  primitive_triplets(b)])
-    for x, y, z in set(result):
-        a, b, c = (2 * x, 2 * y, 2 * z)
-        while c <= max:
-            if a >= min:
-                result.add((a, b, c))
-            a, b, c = (a + x, b + y, c + z)
-    return set([(d, e, f) for d, e, f in result if d >= min and f <= max])
+from math import sqrt, ceil
 
 
-def maxMin(a, b):
-    return (a, b) if a > b else (b, a)
+def triplets_in_range(min, max):
+    result = []
+    for b in range(4, max + 1, 4):
+        for x, y, z in primitive_triplets(b):
+            a, b, c = (x, y, z)
+            while a < min:
+                a, b, c = (a + x, b + y, c + z)
+            while c <= max:
+                result.append((a, b, c))
+                a, b, c = (a + x, b + y, c + z)
+    return result
 
 
 def gcd(x, y):
+    """gcd(x, 0) = x and gcd(x, y) = gcd(y, x % y).
+    See Euclidean Algorithm
+    (https://en.wikipedia.org/wiki/Greatest_common_divisor#Using_Euclid's_algorithm)
+    for more information
+    """
     while (y != 0):
         x, y = y, x % y
     return x
 
 
 def primitive_triplets(b):
```

Comment:
`m = mn // n` will always guarantee `m > n`, so no sort required there.

The second sort can be replaced with the following:

```Python
a = m2 - n2
b = 2 * m * n
if a > b:
    a, b = b, a
yield a, b, m2 + n2
```

On my machine, this is ~25% faster than the comprehension solution. And yes, I would agree it is more readable.

Reply:
After further testing, a modification of your implementation without the extra function call does not perform noticeably better than the 2-function form, so that's good as-is.

Reply:
Your right, the second sort can be replaced with that ... but honestly I'd just keep the second sort it, it's more readable, and under the hood it's doing essentially the same thing.